### ISSUE TITLE
Rarefaction

### ISSUE THREAD

**Issue by philipbaileynar** [namc-utah/namc-bug-db#83]
From @Feldspar5 by email on 4 Mar 2021:

As part of the benthic modeling process, the original list of IDed organisms is rarified to a fixed count (e.g. 300 or 400).  This list of bugs is then passed into the model run.  Any subsequent attempts to run that same sample through the model usually yields different results, because of the randomized rarefaction process.  I cannot remember if we have discussed whether we will be storing that rarefied list, or if it will enter a temporary table as it currently does.  I am not certain it matters as model output will be stored, which means we would never need to rerun a sample-model combination but simply retrieve results.

**Comment by philipbaileynar** [comment]
The plan is to store sufficient information from the rarefaction process such that we can always reproduce the same metrics.

Question: the `sample.organisms` table that replaces `PilotDB.BugData` stores multiple animals per record. Please confirm that rarefaction needs to select 300 or 400 animals and not 300 or 400 database records?

Example: If a sample has two records in BugData with a `SplitCount` of 250 each and we are looking for a rarefaction of 300 then the process first needs to "expand" the two BugData records into 500 animals and from that randomly select 300.

**Comment by Feldspar5** [comment]
I see the confusion, and your understanding seems correct.  Lab IDs 500 individuals, with only 2 taxa.  These are 2 records in the db.  Rarefaction is a withdrawal without replacement of the fixed count number of animals for the given model (300, 400).  This new rarified list are the data used in the modeling process.  This reduces bias introduced as one lab overzealously IDs 1200 animals (leading to increased taxa richness), versus another lab IDing 600 organisms.  We simply 'resample' the sample so both samples are on even numerical footing.

**Comment by philipbaileynar** [comment]
Thanks!

More questions:

Question 1:
Is the rarefaction number (300, 400) considered a property of each model? In other words model 'UT_DEQ_2015' always requires 300 and `WY2018_NE PLAINS` always 400 (fake numbers)? If so, then this is a property that I need to store with models

Question 2:
What happens when a sample has fewer animals than the rarefaction number required by a model?

**Comment by Feldspar5** [comment]
1. Yes.  A model property includes the Fixed Count.  I can provide these to you.
2. Model is run normally.  Depending on model, results are often flagged as something like 'interpret with caution'.  This could trigger a revisit rather than an assessment of 'poor'

**Comment by courtwrightj01** [comment]
Trip can you provide the other model fields we want added to that table as well

**Comment by Feldspar5** [comment]
Yes, I am going to continue appending fields to the sheet I shared with Alex yesterday.

**Comment by Feldspar5** [comment]
Do we need to provide NAR with conditions/thresholds for all models?  For almost every script we have this is currently calculated post-R, a process we should roll into the R scripts themselves.  If we do that, then Phillip likely does not need them at this point.

**Comment by courtwrightj01** [comment]
Should be included in the table but the scripts should handle the condition assignments

**Comment by philipbaileynar** [comment]
If the conditions and logic for thresholding are 100% completely consistent across models then we can consider putting the thresholds in the database and perhaps even performing the classification there after R processing is complete.

1. If there is even a single exception or special treatment between models then the thresholds should be kept in R.
2. If the thresholds will change over time (other than correcting for errors) then I don't recommend putting them in the database.

Here's a typical thresholding table design that might work:

|Model|Threshold Label|MinValue|MaxValue|
|---|---|---|---|
|Model 1|Poor||0.2|
|Model 1|Fair|0.2|0.7|
|Model 1|Fair|0.7||
|Model 2|Terrible||12.34|
|Model 2|Not Too Bad|12.34|14|
|Model 2|OK|14|20.5|
|Model 2|Fantastic|20.5||

* The code that applies these values can be coded with the appropriate "less than" or "less than or equals" logic.
* Note how NULLs are used to express the first and last bins of each classification.

I do not feel strongly about putting the thresholds in the database, but this is one way it could be done.

**Comment by Feldspar5** [comment]
Gotta leave in R then.  Some have 3 conditions, others 4.  And the textual statements differ wildly too.

**Comment by philipbaileynar** [comment]
Number of conditions is easy. Different textual statements is a deal breaker. Keep in R.

**Comment by philipbaileynar** [comment]
A correction before I close this issue:

We are going to be able to reproduce reports. But this is because we will be storing metrics for reports. Not because we are storing the rarefaction parameters necessary to reproduce the sub-sample list of animals used to calculate metrics.

**Comment by courtwrightj01** [comment]
Seems like we want to store the rarefaction parameters or at least just the raw subsample list of animals prior to metrics being calculated so we can reuse those same animals in multiple ways. Trip you agree?

**Comment by Feldspar5** [comment]
I think that is what we want - without that list of rarified taxa, we can never perfectly recreate new metrics.  Does storing these data create a new table?

**Comment by philipbaileynar** [comment]
Glad I asked!

We can store the rarified taxa. It's another table and more data, but it's technically not hard.

To clarify, using my example above with a sample of two BugData records:

* 250 animals of taxa A and life stage B
* 250 animals of taxa C and life stage D

Running a model with a fixed count of 300 would produce two new records in the new table (numbers are fake examples):

* 137 animals of taxa A and life stage B
* 163 animals of taxa C and life stage D

Does that make sense?

**Comment by Feldspar5** [comment]
That is correct (although we do not use lifestage in our models, but fine to keep that field too).  These new records would then be associated with a date, model run?

**Comment by philipbaileynar** [comment]
Yes, @Feldspar5 a date and model run. Ideally we would also track some kind of model run source code version number. This will require some discipline for NAMC staff to track their code in GitHub with official version numbers. It's easy and we can teach the process. Alternatively, we simply rely on the date of the model run and cross reference it to the state of GitHub repo at the time the model was run.

So rarefaction is only based on taxa? bug size is also ignored? Is there any other stratification (e.g. BugData notes? big rare count etc)?

**Comment by Feldspar5** [comment]
Routine only uses taxon name.  Data (taxa names) are first standardized/harmonized to OTUs, then rarified.  Big Rare are never used.

**Comment by philipbaileynar** [comment]
What is the structure of the rarified data passed to the model?

For example, if a model has a fixed count of 300 is the data like this:

|scientific name|quantity|
|---|---|
|taxa1|200|
|taxa2|50|
|taxa3|50|

or just a straight list of 300 taxa names (some of which might repeat)?

**Comment by Feldspar5** [comment]
I cannot speak to all rarefied data, but most of the output I produce and feed into a model end up as site X taxon matrix populated with counts.  Above example would be for single site.  This usually starts as a list that gets pivoted and then fed into model (but not all models require same input format).  For example, CDPHE model data are rarefied in the model black box.
